runtime/master/src/main/java/edu/snu/nemo/runtime/master/scheduler/SourceLocationAwareSchedulingPolicy.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.reef.annotations.audience.DriverSide;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+import java.util.*;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This policy is same as {@link RoundRobinSchedulingPolicy}, however for TaskGroups
+ * with {@link edu.snu.nemo.common.ir.vertex.SourceVertex}, it tries to pick one of the executors
+ * where the corresponding data resides.
+ */
+@ThreadSafe
+@DriverSide
+public final class SourceLocationAwareSchedulingPolicy implements SchedulingPolicy {
+  private static final Logger LOG = LoggerFactory.getLogger(SourceLocationAwareSchedulingPolicy.class);
+
+  private final ExecutorRegistry executorRegistry;
+  private final RoundRobinSchedulingPolicy roundRobinSchedulingPolicy;
+  private final long scheduleTimeoutMs;
+
+  @Inject
+  private SourceLocationAwareSchedulingPolicy(final ExecutorRegistry executorRegistry,
+                                              final RoundRobinSchedulingPolicy roundRobinSchedulingPolicy) {
+    this.executorRegistry = executorRegistry;
+    this.roundRobinSchedulingPolicy = roundRobinSchedulingPolicy;
+    this.scheduleTimeoutMs = roundRobinSchedulingPolicy.getScheduleTimeoutMs();
+  }
+
+  @Override
+  public long getScheduleTimeoutMs() {
+    return scheduleTimeoutMs;
+  }
+
+  /**
+   * Try to schedule a TaskGroup.
+   * If the TaskGroup has one or more source tasks, this method schedules the task group to one of the physical nodes,
+   * chosen from union of set of locations where splits of each source task resides.
+   * If the TaskGroup has no source tasks, falls back to {@link RoundRobinSchedulingPolicy}.
+   * @param scheduledTaskGroup to schedule.
+   * @param jobStateManager jobStateManager which the TaskGroup belongs to.
+   * @return true if the task group is successfully scheduled, false otherwise.
+   */
+  @Override
+  public synchronized boolean scheduleTaskGroup(final ScheduledTaskGroup scheduledTaskGroup,
+                                                final JobStateManager jobStateManager) {
+    Set<String> sourceLocations = Collections.emptySet();
+    try {
+      sourceLocations = getSourceLocations(scheduledTaskGroup.getLogicalTaskIdToReadable().values());
+    } catch (final Exception e) {
+      LOG.warn(String.format("Exception while trying to get source location for %s",
+          scheduledTaskGroup.getTaskGroupId()), e);
+    }
+    if (sourceLocations.size() == 0) {
+      // No source location information found, fall back to the RoundRobinSchedulingPolicy
+      return roundRobinSchedulingPolicy.scheduleTaskGroup(scheduledTaskGroup, jobStateManager);
+    }
+
+    if (scheduleToLocalNode(scheduledTaskGroup, jobStateManager, sourceLocations)) {
+      return true;
+    } else {
+      try {
+        Thread.sleep(scheduleTimeoutMs);
```

Comment:
Can we wake up this thread like `RoundRobinSchedulingPolicy#signalPossiblyWaitingScheduler`?

Reply:
Yup. Thanks.